### PR TITLE
Add `cumprod()` function for cumulative product calculation

### PR DESCRIPTION
**Pull request by samukweku**
- add `cumprod()` function for cumulative product calculation;
- refactor `cumsum()` internals to also support `cumprod()`;

WIP for #3279

### CODE REVIEW

File: src/core/column/cumsumprod.h [hunk 1]
```diff
@@ -28,14 +28,15 @@
 namespace dt {
 
 
-template <typename T>
-class Cumsum_ColumnImpl : public Virtual_ColumnImpl {
+template <typename T, bool MIN>
```
Comment (oleksiyskononenko) [inline src/core/column/cumsumprod.h:31]: Better to call the template parameter not `MIN`, but `SUM`.

File: src/core/column/cumsumprod.h [hunk 2]
```diff
@@ -57,11 +58,20 @@ class Cumsum_ColumnImpl : public Virtual_ColumnImpl {
 
           T val;
           bool is_valid = col_.get_element(i1, &val);
+          if (MIN){
           data[i1] = is_valid? val : 0;
-
+          }
```
Comment (oleksiyskononenko) [inline src/core/column/cumsumprod.h:63]: Need to adjust indentation for line 62.

File: src/core/column/cumsumprod.h [hunk 3]
```diff
@@ -57,11 +58,20 @@ class Cumsum_ColumnImpl : public Virtual_ColumnImpl {
...
+          else {
+            data[i1] = is_valid? val : 1;
+          }
           for (size_t i = i1 + 1; i < i2; ++i) {
             is_valid = col_.get_element(i, &val);
+            if (MIN) {
             data[i] = data[i - 1] + (is_valid? val : 0);
+            }
```
Comment (oleksiyskononenko) [inline src/core/column/cumsumprod.h:71]: Need to adjust indentation for line 70.

File: src/core/expr/fexpr_cumsumprod.cc [hunk 4]
```diff
@@ -33,16 +33,18 @@
 namespace dt {
 namespace expr {
 
-class FExpr_cumsum : public FExpr_Func {
+template <bool MIN>
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_cumsumprod.cc:36]: `MIN` -> `SUM`

File: src/core/expr/fexpr_cumsumprod.cc [hunk 5]
```diff
@@ -33,16 +33,18 @@
...
     ptrExpr arg_;
 
   public:
-    FExpr_cumsum(ptrExpr&& arg)
+    FExpr_CumSumProd(ptrExpr&& arg)
       : arg_(std::move(arg)) {}
 
     std::string repr() const override{
-      std::string out = "cumsum(";
+      std::string out = MIN ? "cumsum" : "cumprod";
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_cumsumprod.cc:46]:
```suggestion
      std::string out = SUM? "cumsum" : "cumprod";
```

File: src/core/expr/fexpr_cumsumprod.cc [hunk 6]
```diff
@@ -69,9 +71,12 @@ class FExpr_cumsum : public FExpr_Func {
...
       SType stype = col.stype();
       switch (stype) {
-        case SType::VOID:    return Column(new ConstInt_ColumnImpl(
+        case SType::VOID:    if (MIN) {return Column(new ConstInt_ColumnImpl(
                                       col.nrows(), 0, SType::INT64
-                                    ));
+                                    ));}
+                             else {return Column(new ConstInt_ColumnImpl(
+                                      col.nrows(), 1, SType::INT64
+                                    ));}
```
Comment (oleksiyskononenko) [inline src/core/expr/fexpr_cumsumprod.cc:79]: No need for brackets when there is just one line of code. Removing them will improve readability.

Comment (oleksiyskononenko) [reply src/core/expr/fexpr_cumsumprod.cc]: The brackets around the return statement. If you want to keep them, it is better to move the return to the next line. Or simply remove them.

File: src/core/column/cumsumprod.h [hunk 7]
```diff
@@ -25,68 +25,84 @@
...
 
   public:
-    Cumsum_ColumnImpl(Column&& col, const Groupby& gby)
-      : Virtual_ColumnImpl(col.nrows(), col.stype()),
-        col_(std::move(col)),
-        gby_(gby)
+    CumSumProd_ColumnImpl(Column &&col, const Groupby &gby)
+        : Virtual_ColumnImpl(col.nrows(), col.stype()),
+          col_(std::move(col)),
+          gby_(gby)
```
[the remark this answers is not in the document]
Comment (samukweku) [reply src/core/column/cumsumprod.h:43]: Is there a C++ formatter? I  use black for all my formatting, it would be nice to have a dedicated formatter for C++

File: docs/api/dt/cumprod.rst [hunk 8]
```diff
@@ -0,0 +1,89 @@
+
+.. xfunction:: datatable.cumprod
+    :src: src/core/expr/fexpr_cumprod.cc pyfn_cumprod
```
Comment (oleksiyskononenko) [inline docs/api/dt/cumprod.rst:3]:
```suggestion
    :src: src/core/expr/fexpr_cumsumprod.cc pyfn_cumprod
```

File: ci/Jenkinsfile.groovy [hunk 9]
```diff
@@ -631,9 +631,9 @@ ansiColor('xterm') {
                                 findFiles(glob: "*").each {
                                     def sha256 = sh(script: """python -c "import hashlib;print(hashlib.sha256(open('${it.name}', 'rb').read()).hexdigest())" """,
                                                     returnStdout: true).trim()
-                                    def s3path = "dev/datatable-${versionText}/${it.name}"
-                                    s3cmd += "s3cmd put -P dist/${it.name} ${S3_BASE}/${s3path}\n"
-                                    pyindex_links += """  <li><a href="${S3_URL}/${encodeURL(s3path)}#sha256=${sha256}">${it.name}</a></li>\n"""
+                                    def s3path = "/dev/datatable-${versionText}/"
+                                    s3cmd += "s3cmd put -P dist/${it.name} ${S3_BASE}${s3path}${it.name}\n"
+                                    pyindex_links += """  <li><a href="${S3_URL}${s3path}${encodeURL(it.name)}#sha256=${sha256}">${it.name}</a></li>\n"""
```
Comment (oleksiyskononenko) [inline ci/Jenkinsfile.groovy:636]: Could you please `git merge main` onto this branch? Not sure why these changes are shown by github since they're already in `main`.

File: docs/api/dt/cumsum.rst [hunk 10]
```diff
@@ -30,7 +30,7 @@
 
     Create a sample datatable frame::
 
-        >>> from datatable import dt, f
+        >>> from datatable import dt, f, by
```
Comment (oleksiyskononenko) [inline docs/api/dt/cumsum.rst:33]: You need to update path on line 3 in this file.

File: src/core/column/cumsumprod.h [hunk 11]
```diff
@@ -25,68 +25,78 @@
...
 
       col_out = std::move(col);
     }
 
-
-    ColumnImpl* clone() const override {
-      return new Cumsum_ColumnImpl(Column(col_), gby_);
+    ColumnImpl *clone() const override
+    {
+      return new CumSumProd_ColumnImpl(Column(col_), gby_);
```
Comment (oleksiyskononenko) [inline src/core/column/cumsumprod.h:84]: So I'm not sure what has happened to formatting, but it seems to me that you have some automatic tool that reformatted the code. We normally keep `{` on the same line as the function/class name.

**Comment by oleksiyskononenko** [comment]
Build failed for some timeout reasons not related to this PR. I've re-triggered the build, let's see if it is green this time.

File: src/core/column/cumsumprod.h (hunk 11 above)
Comment (samukweku) [reply src/core/column/cumsumprod.h:84]: no worries; I'll see how to avoid formatting when working on the datatable repo locally
